File: Shopii/app/src/main/java/com/example/shopii/services/ProductService.java
package com.example.shopii.services;

import com.example.shopii.models.Product;
import com.example.shopii.models.ProductCategory;

import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Query;
import java.util.List;

public interface ProductService {
    @GET("products")
    Call<List<Product>> getProducts();

    @GET("products")
    Call<List<Product>> getProductsByCategory(@Query("category") ProductCategory category);

    @GET("products")
    Call<List<Product>> getProductsByPriceRange(@Query("minPrice") double minPrice, @Query("maxPrice") double maxPrice);

    @GET("products")
    Call<List<Product>> getProductsByRating(@Query("minRating") double minRating);
} 

File: ShopiiAPI/APIService/Services/ProductService.cs
using APIService.Models;
using APIService.Repositories;

namespace APIService.Services
{
    public class ProductService : IProductService
    {
        private readonly IProductRepository _productRepository;

        public ProductService(IProductRepository productRepository)
        {
            _productRepository = productRepository;
        }

        public async Task<IEnumerable<Product>> GetProductsAsync()
        {
            return await _productRepository.GetAllAsync();
        }

        public async Task<IEnumerable<Product>> GetProductsByCategoryAsync(ProductCategory category)
        {
            return await _productRepository.GetProductsByCategoryAsync(category);
        }

        public async Task<IEnumerable<Product>> GetProductsByPriceRangeAsync(double minPrice, double maxPrice)
        {
            return await _productRepository.GetProductsByPriceRangeAsync(minPrice, maxPrice);
        }

        public async Task<IEnumerable<Product>> GetProductsByRatingAsync(double minRating)
        {
            return await _productRepository.GetProductsByRatingAsync(minRating);
        }
    }

    public interface IProductService
    {
        Task<IEnumerable<Product>> GetProductsAsync();
        Task<IEnumerable<Product>> GetProductsByCategoryAsync(ProductCategory category);
        Task<IEnumerable<Product>> GetProductsByPriceRangeAsync(double minPrice, double maxPrice);
        Task<IEnumerable<Product>> GetProductsByRatingAsync(double minRating);
    }
} 